dipy/reconst/fwdti.py

```diff
@@ -0,0 +1,435 @@
+
+
+def _wls_iter(design_matrix, inv_design, sig, min_diffusivity, Diso=3e-3,
+              piterations=3, riterations=2):
+    """ Helper function used by wls_fit_tensor - Applies WLS fit of the
+    water free elimination model to single voxel signals.
+
+    Parameters
+    ----------
+    design_matrix : array (g, 7)
+        Design matrix holding the covariants used to solve for the regression
+        coefficients
+    inv_design : array (g, 7)
+        Inverse of the design matrix.
+    sig : array (g, )
+        Diffusion-weighted signal for a single voxel data.
+    min_diffusivity : float
+        Because negative eigenvalues are not physical and small eigenvalues,
+        much smaller than the diffusion weighting, cause quite a lot of noise
+        in metrics such as fa, diffusivity values smaller than
+        `min_diffusivity` are replaced with `min_diffusivity`.
+    Diso : float, optional
+        Value of the free water isotropic diffusion. Default is set to 3e-3
+        $mm^{2}.s^{-1}$. Please ajust this value if you are assuming different
+        units of diffusion.
+    piterations : inter, optional
+        Number of iterations used to refine the precision of f. Default is set
+        to 3 corresponding to a precision of 0.01.
+    riterations : inter, optional
+        Number of iteration repetitions with adapted S0. To insure that S0 is
+        taken as a model free parameter Each precision iteration is repeated
+        riterations times with adapted S0.
+
+    Returns
+    -------
+    All parameters estimated from the free water tensor model.
+    Parameters are ordered as follows:
+        1) Three diffusion tensor's eigenvalues
+        2) Three lines of the eigenvector matrix each containing the
+           first, second and third coordinates of the eigenvector
+        3) The volume fraction of the free water compartment
+    """
+    W = design_matrix
+
+    # DTI ordinary linear least square solution
+    log_s = np.log(sig)
+    ols_result = np.dot(inv_design, log_s)
+
+    # Define weights as diag(yn**2)
+    S2 = np.diag(np.exp(2 * np.dot(W, ols_result)))
+
+    # DTI weighted linear least square solution
+    WTS2 = np.dot(W.T, S2)
+    inv_WT_S2_W = np.linalg.pinv(np.dot(WTS2, W))
+    WT_S2_LS = np.dot(WTS2, log_s)
+    params = np.dot(inv_WT_S2_W, WT_S2_LS)
+    
+    # General free-water signal contribution
+    fwsig = np.exp(np.dot(design_matrix, 
+                          np.array([Diso, 0, Diso, 0, 0, Diso, -np.log(1.)])))
+
+    df = 1  # initialize precision
+    flow = 0  # lower f evaluated
+    fhig = 1  # higher f evaluated
+    ns = 9  # initial number of samples per iteration
+    nvol = len(sig)
+    for p in range(piterations):
+        df = df * 0.1
+        fs = np.linspace(flow+df, fhig-df, num=ns)  # sampling f
+        # repeat fw contribution for all the samples
+        SFW = np.array([fwsig,]*ns)
+        FS, SI = np.meshgrid(fs, sig)
+        for r in range(riterations):
```

Comment:
@samuelstjean - Well, for this to converge to a given precision it will require always the same amount of steps, that is why I didn't use break statements. Users can just decrease the number of steps it they want to decrease the precision - see the function's documentation.